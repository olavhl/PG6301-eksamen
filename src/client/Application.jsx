import React from "react";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";

export function Application() {
  return (
    <BrowserRouter>
      <header>
        <Link to={"/"}>
          <button>Home</button>
        </Link>
      </header>
      <Switch>
        <Route exact path={"/login"}>
          <h1>Login</h1>
        </Route>
        <Route path={"/login/callback"}>
          <h1>Login Callback</h1>
        </Route>
        <Route exact path={"/"}>
          <h1>Welcome to the chatting page</h1>
          <ul>
            <li>
              <Link to={"/login"}>Login</Link>
            </li>
          </ul>
        </Route>
        <Route>
          <h1>Page not found...</h1>
        </Route>
      </Switch>
    </BrowserRouter>
  );
}
