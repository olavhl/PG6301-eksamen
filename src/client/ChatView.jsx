import React, { useState } from "react";

export function ChatView({ onSendMessage, chat }) {
  const [message, setMessage] = useState("");
  // Feedback
  const [feedBack, setFeedBack] = useState("");

  function handleSubmitChat(e) {
    e.preventDefault();

    if (message === "") {
      setFeedBack("Write the message you want to send!");
      return;
    }

    onSendMessage(message);
    setMessage("");
  }

  return (
    <div className={"chat"}>
      <header>
        <h1>Chat</h1>
      </header>
      <main>
        {chat?.map((message, index) => (
          <div key={`${message}__${index}`}>{message}</div>
        ))}
      </main>
      {feedBack && <div className={"feedback"}>{feedBack}</div>}
      <footer>
        <form onSubmit={handleSubmitChat}>
          <input
            type="text"
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button>Send</button>
        </form>
      </footer>
    </div>
  );
}
