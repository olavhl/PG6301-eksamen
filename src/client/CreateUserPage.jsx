import React, { useState } from "react";
import { InputField } from "./components/InputField";
import { useHistory } from "react-router";

export function CreateUserPage({ userApi }) {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  // Feedback
  const [feedBack, setFeedBack] = useState("");

  const history = useHistory();

  async function handleSubmit(e) {
    e.preventDefault();

    if (firstName === "" || lastName === "" || email === "") {
      setFeedBack(
        "You need to implement First and Last name, as well as email"
      );
      return;
    }

    await userApi.createUser({ firstName, lastName, email });
    history.push("/user/list");
  }

  return (
    <div className={"cont"}>
      <h1>Create user</h1>
      <form onSubmit={handleSubmit}>
        <InputField
          value={firstName}
          label={"First name"}
          onChangeValue={setFirstName}
        />
        <InputField
          value={lastName}
          label={"Last name"}
          onChangeValue={setLastName}
        />
        <InputField value={email} label={"Email"} onChangeValue={setEmail} />
        <button>Submit</button>
      </form>
      <br />
      {feedBack && <div className={"feedback"}>{feedBack}</div>}
    </div>
  );
}
