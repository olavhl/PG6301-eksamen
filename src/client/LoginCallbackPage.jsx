import React, { useEffect } from "react";
import { useHistory } from "react-router";

export function LoginCallbackPage({ onAccessToken }) {
  const hash = Object.fromEntries(
    new URLSearchParams(window.location.hash.substring(1))
  );

  const history = useHistory();

  useEffect(() => {
    const { access_token } = hash;
    onAccessToken(access_token);
    history.push("/");
  }, []);

  return <h1>Login Callback</h1>;
}
