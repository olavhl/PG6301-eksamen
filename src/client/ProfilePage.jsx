import React from "react";
import { useLoading } from "./lib/useLoading";
import { ErrorView } from "./components/ErrorView";
import { LoadingView } from "./components/LoadingView";

export function ProfilePage({ loadProfile }) {
  const { data: user, error, loading, reload } = useLoading(
    async () => await loadProfile()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading) {
    return <LoadingView />;
  }

  return (
    <div>
      <h1>Profile Page</h1>
      {user && (
        <div>
          <p>First name: {user.given_name}</p>
          <p>Last name: {user.family_name}</p>
          <p>Email: {user.email}</p>
          <img src={user.picture} alt="Profile picture" />
        </div>
      )}
    </div>
  );
}
