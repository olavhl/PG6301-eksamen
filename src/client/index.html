<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Messaging System</title>
    <link rel="stylesheet" href="style.css" type="text/css">
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
