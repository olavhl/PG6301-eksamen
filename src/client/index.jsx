import React from 'react';
import ReactDOM from 'react-dom';

function Application() {
    return <h1>Hello</h1>;
}

ReactDOM.render(<Application/>, document.getElementById("app"));