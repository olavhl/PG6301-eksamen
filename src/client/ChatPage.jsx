import React, { useEffect, useState } from "react";
import { ChatView } from "./ChatView";

export function ChatPage() {
  const [chat, setChat] = useState([]);
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket("ws://localhost:3000/chat");
    ws.onopen = (event) => {
      console.log("opened", event);
    };

    ws.onmessage = (event) => {
      console.log("message", event);
      setChat((chat) => [...chat, event.data]);
    };

    ws.onclose = (event) => {
      console.log("close", event);
    };

    setWs(ws);
  }, []);

  return <ChatView chat={chat} onSendMessage={(message) => ws.send(message)} />;
}
