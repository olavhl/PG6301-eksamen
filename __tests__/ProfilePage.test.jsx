import React from "react";
import ReactDOM from "react-dom";
import { ProfilePage } from "../src/client/ProfilePage";
import { act } from "react-dom/test-utils";

async function renderToDom(component) {
  const container = document.createElement("div");
  document.body.appendChild(container);
  await act(async () => ReactDOM.render(component, container));

  return container;
}

describe("profile page", () => {
  it("shows loading-screen", async () => {
    const loadProfile = () => new Promise(() => {});

    const container = await renderToDom(
      <ProfilePage loadProfile={loadProfile} />
    );

    expect(container.querySelector("div").textContent).toEqual("Loading...");
    expect(container.innerHTML).toMatchSnapshot();
  });

  it("shows error", async () => {
    const loadProfile = async () => {
      throw Error("Unauthorized");
    };
    const container = await renderToDom(
      <ProfilePage loadProfile={loadProfile} />
    );

    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("div").textContent).toEqual(
      "Something went wrongError: Unauthorized"
    );
  });
});
