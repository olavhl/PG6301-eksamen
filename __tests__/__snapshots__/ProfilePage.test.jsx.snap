// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile page shows error 1`] = `"<div class=\\"cont\\"><div>Something went wrong<br>Error: Unauthorized</div><button>Try again</button></div>"`;

exports[`profile page shows loading-screen 1`] = `"<div class=\\"cont\\">Loading...</div>"`;
